---
title: "BMS"
author: "Yuxi Ke"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    theme: "flatly"
    number_sections: yes
    toc: yes
    toc_float: yes
    code_folding: hide
---

```{r config, warning = FALSE}

# **MODIFY THIS CHUNK**
project_id <- "bms" # determines the name of the cache folder
doc_id     <- file.path("02-2_GEX_per_cell/") # determines name of the subfolder of `figures` where pngs/pdfs are saved
out        <- file.path(here::here("Output/"), doc_id); dir.create(out, recursive = TRUE)
figout     <- file.path(here::here("Fig/"), doc_id); dir.create(figout, recursive = TRUE)
```

```{r setup, include = FALSE}

# NO NEED TO MODIFY THIS CHUNK
here::i_am(".gitignore")
knitr::opts_chunk$set(message = TRUE,
                      warning = FALSE,
                      error = FALSE,
                      cache = FALSE,
                      fig.path = figout,
                      fig.keep = "all",
                      dev = c("png", "pdf"),
                      cache.lazy = FALSE)

grDevices::pdf.options(useDingbats = FALSE)
options(knitr.table.format = "html") 

```

***

# Overview

This file

- loads and computes per cell QC for gene expression data
- computes doublet scores without further processing
(wait for ATAC side per cell doublet scores)
- down samples some over-sequenced samples
Yuxi Ke, 12/02/2024


# Libraries

```{r libs, warning = FALSE, message = FALSE}
library(here)
source(here::here("Scripts/Lib/util.R"))
source(here::here("Scripts/Lib/plotting_Greenleaf.R"))
source(here::here("Scripts/Lib/plotting.R"))

library(dplyr)
library(Seurat)
suppressPackageStartupMessages(library(scDblFinder))
# library(DropletUtils)
library(ggplot2)

ggplot2::theme_set(theme_BOR())

set.seed(100)
```


# Load data

## Define file names

```{r}
# oak.path <- "/oak/stanford/groups/wjg/kyx/data/bms"

gex.obj.path <- file.path(here::here("Data/SeuratObjects/"), "BMS_YK_gex_early.rds")

gex_sample_ls <- c("YK01_En7Apos", "YK02_En7Bneg", "YK04_En7Cneg", 
                "YK05_En7Cpos", "YK06_En7ReAneg", "YK07_En7ReApos", "YK08_En7ReBpos",  
                "YK09_En28Apos", "YK10_En28Cneg", "YK11_En28Cpos", "YK12_En28Dneg", "YK13_En28Dpos", 
                "YK14_En28ReBneg", "YK15_En28ReBpos", "YK16_En28ReCneg", "YK17_En28ReCpos")

gex_oak_dir <- "~/oak/data/bms/mm10_aligned"
gex_data_dir <- lapply(gex_sample_ls, function (x) file.path(gex_oak_dir, x)) %>% unlist() %>% c()
names(gex_data_dir) <- gex_sample_ls
metadata_dir <- here::here("Data/Metadata/Yuxi_sample_sheet.csv")
allen.obj.path <- here::here("Data/SeuratObjects/Allen_PFC_1in20_02-1.rds")
```

## Define colors

```{r BMS color maps}
bms_color = list(
    sample = getColorMap(cmaps_BOR$circus, 16),
    seurat_clusters = getColorMap(cmaps_BOR$iron_man, 40),
    doublet = c("Singlet" = "#8A9FD1", "Doublet" = "#D51F26",
            "singlet" = "#8A9FD1", "doublet" = "#D51F26",
            "2-singlet" = "#8A9FD1", "1-doublet" = "#D51F26")
)
```


## Read data


```{r Read Allen data}
allen <- readRDS(allen.obj.path)
```

```{r Read data by sample}
gex_object_list <- vector(mode = "list", length = length(gex_data_dir))
names(gex_object_list) <- names(gex_data_dir)

for (i in 1:length(gex_object_list)) {
    data_dir <- gex_data_dir[i]
    print("===============")
    print(names(data_dir))
    # Load object
    tmp_seurat_object <- LoadSeuratObject(data_dir, project=names(data_dir))
    gex_object_list[[names(gex_object_list)[[i]]]] <- tmp_seurat_object
}
rm(tmp_seurat_object)
```

```{r}
# Simply read from the data files
saveRDS(gex_object_list, here::here("Data/SeuratObjects/gex_object_list_raw.rds"))
```


```{r}
for (i in 1:length(gex_object_list)) {
    print("====================")
    print(names(gex_object_list)[[i]])
    # Load object
    tmp_seurat_object <- gex_object_list[[names(gex_object_list)[[i]]]]
    
    # Filter out extremely low count cells
    # although per cell QC filter should be after doublet calling
    # Using clusters since homotypic doublets are completely okay for me
    tmp_seurat_object <- subset(tmp_seurat_object, subset = nCount_RNA > 200)
    sce <- scDblFinder(as.SingleCellExperiment(tmp_seurat_object),
                   clusters = TRUE,
                   dbr.sd = 0,
                   dbr.per1k = 0.006)
    tmp_seurat_object <- AddMetaData(
        object = tmp_seurat_object,
        metadata = sce$scDblFinder.class,
        col.name = "scDblFinder_class"
    )
    tmp_seurat_object <- AddMetaData(
        object = tmp_seurat_object,
        metadata = sce$scDblFinder.score,
        col.name = "scDblFinder_score"
    ) 
    gex_object_list[[names(gex_object_list)[[i]]]] <- tmp_seurat_object
}
```

```{r}
# Only with extreme low count cells removed and doublets called
saveRDS(gex_object_list, here::here("Data/SeuratObjects/gex_object_list_doublet.rds"))
```

# Per sample

## Preprocess

```{r Preprocess each sample}
gex_object_list <- lapply(X = gex_object_list, FUN = function(x) {
    x[["percent.mt"]] <- PercentageFeatureSet(x, pattern = "^mt-")
    x <- subset(x, subset = nFeature_RNA > 500 & nFeature_RNA < 7500 & 
                    nCount_RNA > 1000 & nCount_RNA < 8e4 & percent.mt < 5)
    x <- NormalizeData(x, verbose = FALSE)
    x <- FindVariableFeatures(x, selection.method = "mvp", nfeatures = 2000, verbose = F)
    x <- ScaleData(x, verbose = F)
})
```

## Label transfer

```{r kept dims and label col names}
# Define the PCs kept for analysis
keptDims <- 1:40
nDims <- tail(keptDims, 1)

# Define names of the columns to transfer
label.col.names <- c("cluster", "subclass", "neighborhood", "class")
```


```{r Transfer labels per sample}

gex_object_list <- lapply(X = gex_object_list, FUN = function(x) {
    transfer_anchors <- FindTransferAnchors(
        reference = allen,
        query = x,
        dims = keptDims, 
        reference.reduction = "pca"
    )
    
    prediction <- list()
    for (label.col in label.col.names) {
        preds <- TransferData(
            anchorset = transfer_anchors,
            refdata = allen[[]][, paste0(label.col, "_label")]
        )
        preds <- preds[, c(1, ncol(preds))]
        colnames(preds) <- c(label.col, paste0(label.col, ".prediction.score.max"))
        prediction <- append(prediction, preds)
    }
    prediction <- as.data.frame(prediction)
    rownames(prediction) <- Cells(x)
    x <- AddMetaData(x, metadata = prediction)
})


```


```{r}
# Raise error if cell names are empty
stopifnot(! is.null(Cells(gex_object_list[[1]])))
Cells(gex_object_list[[1]])[[1]]
```

```{r}
# with transferred labels
saveRDS(gex_object_list, here::here("Data/SeuratObjects/gex_object_list_labeled.rds"))
```

# Merge samples

```{r Merge to gex object}
gex <- merge(gex_object_list[[1]], y=gex_object_list[2:length(gex_object_list)],
             add.cell.ids = names(gex_object_list),
             project = "PFC")
```


```{r Clean up cell names}
# Replace the second _ to # in cell names
replace_second_underscore <- function(input_string) {
    gsub("^(.*?_.*?)_(.*)$", "\\1#\\2", input_string)
}
new_cell_names <- sapply(Cells(gex), replace_second_underscore)
gex <- RenameCells(gex, new.names = new_cell_names)

# # Rebuild Seurat object because of persisting cell name issues
# #' The `names.delim = "#"` option in `CreateSeuratObject()`
# #' doesn't do what you think it does.
# #' It only changes how cell information is parsed, not how they're made.
# gex_counts <- gex@assays$RNA@counts
# names(colnames(gex_counts)) <- Cells(gex)
# gex <- CreateSeuratObject(counts = gex_counts, 
#                           meta.data = gex@meta.data)
#
# # Extract consistent cell names without attributes
# consistent_cell_names <- unname(colnames(gex@assays$RNA@counts))
# 
# # Force reset rownames of metadata
# rownames(gex@meta.data) <- consistent_cell_names
# 
# # Force reset colnames of count matrix
# colnames(gex@assays$RNA@counts) <- consistent_cell_names
# 
# # Rebuild metadata to remove any attributes
# gex@meta.data <- data.frame(gex@meta.data, row.names = consistent_cell_names)
```

```{r}
saveRDS(gex, here::here("Data/SeuratObjects/gex_merged.rds"))
```


## Process merged object

```{r Process gex}
features <- SelectIntegrationFeatures(object.list = gex_object_list)
rm(gex_object_list)
gex <- NormalizeData(gex, verbose = FALSE)
gex <- ScaleData(gex, verbose = FALSE)
gex <- RunPCA(gex, features = features, verbose = FALSE)

keptDims <- 1:30
gex <- RunUMAP(gex, dims = keptDims, seed.used = 42, verbose = FALSE)
gex <- FindNeighbors(gex, dims = keptDims)
gex <- FindClusters(gex, resolution = 5)
```


# Visualization

## QC per sample

```{r}
VlnPlot(gex, "nCount_RNA", group.by = "orig.ident", pt.size=0) + ylim(c(0,10000))
p <- RidgePlot(gex, "nFeature_RNA", group.by = "orig.ident") 
p
```


## PCs

```{r, fig.height=12, fig.width=12}
DimHeatmap(gex, 1:10, ncol = 4)
```
```{r}
ElbowPlot(gex, 50)
```


## UMAP

```{r}
label.col.names <- c("cluster", "subclass", "neighborhood", "class")

p <- list()
label2plot <- c(label.col.names[c(2,3,4)], "seurat_clusters")
for (label.col in label2plot) {
    n <- nrow(unique(gex[[label.col]]))
    p[[label.col]] <- DimPlot(gex, group.by = label.col, 
                              label = T, repel = TRUE, 
                              label.box = TRUE, label.size = 2,
                              cols = getColorMap(cmaps_BOR$circus, n + 1))+ NoLegend()
}

# Combined plot
gridExtra::grid.arrange(grobs = p)

# Separate plots
p


DimPlot(gex, group.by = "orig.ident", cols = getColorMap(cmaps_BOR$circus, n + 1), shuffle = T)
FeaturePlot(gex, 
            features = c("class.prediction.score.max"),
            cols = c("red", "grey"))
```


```{r}
FeaturePlot(gex, "scDblFinder_score", min.cutoff = 0.9, max.cutoff = 1)
DimPlot(gex, group.by = "scDblFinder_class", cols = bms_color$doublet)
table(gex$scDblFinder_class)
```

## Class markers
```{r}
FeaturePlot(gex, c("Rbfox3", "Snap25", "Slc17a7", "Gad1"))
VlnPlot(gex, features = c("Rbfox3", "Slc17a7", "Gad1"), pt.size = 0,
        group.by = "class", cols = bms_color$seurat_clusters)
```

## Glial markers

```{r}
# Define marker lists
glial_markers <- list(
  GenericGlial = c("Sox9", "Gfap", "Aldh1l1", "Vim"),
  Astrocytes = c("Aqp4", "Gfap", "Slc1a2", "Slc1a3", "Aldh1l1", "Fabp7"),
  Oligodendrocytes = c("Mbp", "Mag", "Plp1", "Cnp", "Olig1", "Olig2"),
  OPCs = c("Pdgfra", "Cspg4", "Sox10", "Nkx2-2"),
  Microglia = c("Cx3cr1", "P2ry12", "Aif1", "Trem2", "Hexb"),
  Ependymal = c("Foxj1", "Vim", "Ttr", "Dnah9")
)

# Add gene scores for each marker set
gex <- AddModuleScore(
  object = gex,
  features = glial_markers,
  name = "GlialType"
)

# Define cell type labels for subtitles
cell_types <- c("Generic Glial", "Astrocytes", "Oligodendrocytes", 
                "OPCs", "Microglia", "Ependymal Cells")

# Create UMAP plots with subtitles
plots <- lapply(1:length(cell_types), function(i) {
  FeaturePlot(gex, features = paste0("GlialType", i), 
              cols = c("lightgrey", "magenta"), reduction = "umap") +
    labs(title = paste(cell_types[i])) +
    theme(plot.title = element_text(hjust = 0.5))
})

# Print all plots
for (p in plots) print(p)

#Combined
gridExtra::grid.arrange(grobs = plots)
```

## QC per cluster

```{r QC Vln plots per cluster}
VlnPlot(gex, c("nCount_RNA", "percent.mt", "scDblFinder_score", "class.prediction.score.max"), pt.size = 0, ncol = 2,
          group.by = "seurat_clusters")
```




```{r}
ggplot(gex[[]], aes(fill=subclass, x=seurat_clusters)) + 
    geom_bar(position = "fill") + coord_flip() + scale_fill_discrete(type=getColorMap(cmaps_BOR$kelly, 26))

ggplot(gex[[]], aes(fill=neighborhood, x=seurat_clusters)) + 
    geom_bar(position = "fill") + coord_flip() + scale_fill_discrete(type=getColorMap(cmaps_BOR$kelly, 26))

ggplot(gex[[]], aes(fill=class, x=seurat_clusters)) + 
    geom_bar(position = "fill") + coord_flip() + scale_fill_discrete(type=getColorMap(cmaps_BOR$stallion, 26))

ggplot(gex[[]], aes(fill=class, x=seurat_clusters)) + 
    geom_bar() + coord_flip() + scale_fill_discrete(type=getColorMap(cmaps_BOR$stallion, 26))

ggplot(gex[[]], aes(x=seurat_clusters, y=class.prediction.score.max)) +
    geom_violin() #+ coord_flip()

pred.score <- gex[[]] %>%
    group_by(seurat_clusters) %>%
    summarize(median = median(class.prediction.score.max),
              p25 = quantile(class.prediction.score.max, 0.25),
              size = length(class.prediction.score.max)) %>%
    arrange(median)

barplot(pred.score$median,
        names.arg = unique(gex$seurat_clusters))
barplot(pred.score$p25,
        names.arg = unique(gex$seurat_clusters))
print(head(pred.score, 15))
```
```{r}
RidgePlot(
    gex,
    "scDblFinder_score",
    group.by = "orig.ident",
)
```
```{r}
RidgePlot(gex, c("nCount_RNA", "nFeature_RNA", "percent.mt"), group.by = "seurat_clusters")
```

# Ambiguous cells

```{r}
# Add ambiguous_cluster to metadata
gex$ambiguous_cluster <- ifelse(
  gex$class.prediction.score.max < 0.5, "Ambiguous", "Not Ambiguous"
)

# Confirm the new column in metadata
table(gex$ambiguous_cluster)

# Subset to ambiguous cells for visualization or analysis
ambiguous_cells <- subset(gex, subset = ambiguous_cluster == "Ambiguous")

# Check marker gene expression
FeaturePlot(ambiguous_cells, features = c("Snap25", "Slc17a7", "Gad1", "Aldh1l1", "Mbp", "Fos", "Hspa1a"))

# Run FindMarkers for ambiguous vs. not ambiguous cells
de_results <- FindMarkers(
  gex,
  ident.1 = "Ambiguous",
  ident.2 = "Not Ambiguous",
  group.by = "ambiguous_cluster"
)

# View the results
head(de_results,50)

# Visualize metadata correlations
FeaturePlot(ambiguous_cells, features = c("percent.mt", "nCount_RNA"))
```
```{r}
FeaturePlot(gex, features = c("Pbx3", "Ptprd", "Csmd1", "Kcnip4", "Meis2", "Rims1"))

```
```{r}
RidgePlot(gex, features = c("neighborhood.prediction.score.max",
                            "class.prediction.score.max"),
          group.by = "orig.ident")
```


# Secondary filters

## Filter
```{r}
subgex <- subset(gex, subset = scDblFinder_class == "singlet")
```

```{r}
non_neuronal_clusters <- gex[[]] %>%
    dplyr::group_by(seurat_clusters) %>%
    dplyr::summarise(
        non_neuronal_prop = mean(class == "Non-Neuronal"),
        .groups = "drop"
    ) %>%
    dplyr::filter(non_neuronal_prop > 0.6) %>%
    dplyr::pull(seurat_clusters)


doublet_clusters <- gex[[]] %>%
    dplyr::group_by(seurat_clusters) %>%
    dplyr::summarise(
        doublet_median = median(scDblFinder_score),
        n_cell = length(scDblFinder_score),
        .groups = "drop"
    ) %>%
    filter(doublet_median > 0.75 & n_cell < 100) %>%
    pull(seurat_clusters)

# low_score_clusters <- gex[[]] %>%
#     group_by(seurat_clusters) %>%
#     summarise(
#         median_prediction_score = median(class.prediction.score.max, na.rm = TRUE),
#         .groups = "drop"
#     ) %>%
#     filter(median_prediction_score < 0.7) %>%
#     pull(seurat_clusters)

`%ni%` <- Negate(`%in%`)
# subgex <- subset(subgex, subset = (seurat_clusters %ni% clusters_to_remove))
# subgex <- subset(subgex, subset = (class %in% c("Glutamatergic", "GABAergic")))
# subgex <- subset(subgex, subset = (subclass %ni% c("Astro", "Oligo")))
```

```{r}
table(gex$class.prediction.score.max > 0.4)

# subgex <- subset(subgex, subset = class.prediction.score.max > 0.6)
```

```{r}
is_singlet <- (gex$scDblFinder_class == "singlet") &
    (gex$seurat_clusters %ni% doublet_clusters)

# Many of the predicted non-neuronal cells are actually just low score cells
# We keep them for now and only remove the clear non-neuronal clusters
# We don't have enough of these cells to come to any statistically-significant 
# conclusions anyways
is_neuron <- (gex$seurat_clusters %ni% non_neuronal_clusters) &
    gex$class.prediction.score.max > 0.4
    # (gex$class %in% c("Glutamatergic", "GABAergic")) &
    # (gex$subclass %ni% c("Astro", "Oligo"))
table(is_singlet)
table(is_neuron)
table(is_singlet & is_neuron)
```
```{r}
gex <- AddMetaData(gex, is_neuron, "is_neuron")
gex <- AddMetaData(gex, is_singlet, "is_singlet")
```

```{r}
saveRDS(gex, here::here("Data/SeuratObjects/gex_tagged.rds"))
```


```{r}
DimPlot(subgex, group.by = "subclass", cols = bms_color$seurat_clusters)
# DimPlot(subgex, group.by = "subclass", cols = bms_color$seurat_clusters,
#         split.by = "orig.ident", ncol = 4) + NoLegend()
```
```{r}
FeaturePlot(subgex, c("neighborhood.prediction.score.max",
                      "class.prediction.score.max"),
            cols = c("deeppink", "grey"))
```

## Re-UMAP

```{r}
subgex <- NormalizeData(subgex)
subgex <- ScaleData(subgex, verbose = FALSE)
subgex <- FindVariableFeatures(subgex)
subgex <- RunPCA(subgex, verbose = FALSE)
```

```{r}
# ElbowPlot(subgex, ndims = 50)
# DimHeatmap(subgex, dims = 16:19)
```

```{r}
keptDimsLess <- 1:30
subgex <- RunUMAP(subgex, dims = keptDimsLess, verbose = FALSE)
subgex <- FindNeighbors(subgex, dims = keptDimsLess)
subgex <- FindClusters(subgex, resolution = 1)
```

```{r}
label.col.names <- c("cluster", "subclass", "neighborhood", "class")

p <- list()
label2plot <- c(label.col.names[c(2,3,4)], "seurat_clusters")
for (label.col in label2plot) {
    n <- nrow(unique(subgex[[label.col]]))
    p[[label.col]] <- DimPlot(subgex, group.by = label.col, 
                              label = T, repel = TRUE, 
                              label.box = TRUE, label.size = 2,
                              cols = getColorMap(cmaps_BOR$circus, n + 1))+ NoLegend()
}

# Combined plot
gridExtra::grid.arrange(grobs = p)

# Separate plots
p

```


```{r}
VlnPlot(subgex, "nCount_RNA", group.by = "seurat_clusters", pt.size=0) + ylim(c(0,10000))
VlnPlot(subgex, "nFeature_RNA", group.by = "seurat_clusters", pt.size=0)
```

```{r}
saveRDS(subgex, here::here("Data/SeuratObjects/subgex.rds"))
```


```{r}

```


```{r}
transferLabels <- function(x){
    transfer_anchors <- FindTransferAnchors(
        reference = allen,
        query = x,
        dims = keptDims, 
        reference.reduction = "pca"
    )
    
    prediction <- list()
    for (label.col in label.col.names) {
        preds <- TransferData(
            anchorset = transfer_anchors,
            refdata = allen[[]][, paste0(label.col, "_label")]
        )
        preds <- preds[, c(1, ncol(preds))]
        colnames(preds) <- c(label.col, paste0(label.col, ".prediction.score.max"))
        prediction <- append(prediction, preds)
    }
    prediction <- as.data.frame(prediction)
    rownames(prediction) <- Cells(x)
    x <- AddMetaData(x, metadata = prediction)
}
```



```{r}
DimPlot(singex, group.by = "seurat_clusters", cols = bms_color$seurat_clusters, label=T)
```
```{r}
# Run FindMarkers for ambiguous vs. not ambiguous cells
de_results <- FindMarkers(
  singex,
  ident.1 = c("15"),
  group.by = "seurat_clusters"
)

head(de_results, 50)
```

```{r}
meta.data <- LoadAirtableSampleMetadata(here::here("Data/Metadata/YK_sample_sheet.csv"))
names(meta.data)[names(meta.data) == "CombinedID"] <- "orig.ident"
cols2keep <- c("orig.ident", "SampleID", "Experiment", "Group", "Gating", "Condition")
joined_meta_data <- left_join(singex[[]], meta.data[, cols2keep], by = "orig.ident")
rownames(joined_meta_data) <- rownames(singex[[]])
singex@meta.data <- joined_meta_data
```

```{r}
DimPlot(singex, group.by = "Group", shuffle = T,
        cols = bms_color$sample)
DimPlot(singex, split.by = "Group", 
        group.by = "Gating", shuffle = T,
        ncol = 2,
        cols = bms_color$sample)
```

```{r}
ggplot(singex[[]], aes(fill=subclass, x=Condition)) + 
    geom_bar(position = "fill") + coord_flip() + scale_fill_discrete(type=getColorMap(cmaps_BOR$kelly, 26))


ggplot(singex[[]], aes(fill=class, x=seurat_clusters)) + 
    geom_bar(position = "fill") + coord_flip() + scale_fill_discrete(type=getColorMap(cmaps_BOR$kelly, 26))
```

```{r}
ambiguous_clusters <- c("5", "11", "17", "18")
singex$ambiguous_cluster <- ifelse(
  singex$seurat_clusters %in% ambiguous_clusters, "Ambiguous", "Not Ambiguous"
)

ggplot(singex[[]], aes(fill=ambiguous_cluster, x=SampleID)) + 
    geom_bar(position = "fill") + coord_flip() + scale_fill_discrete(type=getColorMap(cmaps_BOR$kelly, 26))
```


```{r}
saveRDS(singex, here::here("Data/SeuratObjects/singex.rds"))
```

```{r}
saveRDS(gex, here::here("Data/SeuratObjects/gex_raw.rds"))
```

# Save GEX

```{r Save Seurat object}
saveRDS(gex, gex.obj.path)
```



<!-- # UMAP sanity check -->

<!-- ```{r, fig.width=12, fig.height=24} -->
<!-- DimHeatmap(gex, dims = 30:50,  -->
<!--            ncol = 4, cells = 1000) -->
<!-- ``` -->



<!-- ```{r Sequencing depth on UMAP} -->
<!-- FeaturePlot(gex, -->
<!--             features = c("nCount_RNA", "nFeature_RNA"), -->
<!--             cols = c("gray90", "deeppink") -->
<!--             ) -->
<!-- ``` -->


<!-- # Doublet filtering -->

<!-- ## Visualize across doublet thresholds -->

<!-- ```{r doublet on UMAP} -->
<!-- DimPlot(gex, group.by = "seurat_clusters", cols = bms_color$seurat_clusters) -->

<!-- p.class <- list() -->
<!-- p.score <- list() -->
<!-- # `thresholds` defined earlier in "Read data" -->
<!-- for (thresh in thresholds) { -->
<!--     thresh_str <- as.character(thresh) -->
<!--     class_lbl <- paste0("scDblFinder_class_", thresh_str) -->
<!--     score_lbl <- paste0("scDblFinder_score_", thresh_str) -->

<!--     p.class[[thresh_str]] <- DimPlot(gex, -->
<!--                                      group.by = class_lbl, -->
<!--                                      cols = bms_color$doublet) -->
<!--     p.score[[thresh_str]] <- FeaturePlot(gex, -->
<!--         score_lbl, -->
<!--         cols = c("grey90",bms_color$doublet[["doublet"]]), -->
<!--         min.cutoff = .5) -->
<!-- } -->

<!-- print(p.class) -->
<!-- print(p.score) -->
<!-- ``` -->


<!-- ```{r Percent doublets, fig.width=2} -->
<!-- thresholds <- c(0.001, 0.002, 0.005, 0.01) -->
<!-- doublet_proportions <- c() -->
<!-- for (rate in thresholds){ -->
<!--     doublet_proportions[as.character(rate)] <- 100 * mean(gex[[paste0("scDblFinder_class_", as.character(rate))]] == "doublet") -->
<!-- } -->


<!-- barplot(doublet_proportions, -->
<!--       names.arg = names(doublet_proportions), -->
<!--       xlab = "Doublet Threshold", -->
<!--       ylab = "Percentage of Doublets", -->
<!--       main = "Doublet Detection Proportions", -->
<!--       width = .2, ylim = c(0,5)) + theme_BOR() -->
<!-- ``` -->

<!-- ## Threshold picked -->
<!-- We pick threshold 0.002 -->

<!-- ```{r Doublet score on UMAP split by sample} -->
<!-- FeaturePlot(gex, -->
<!--             "scDblFinder_score_0.002", -->
<!--             split.by = "orig.ident", -->
<!--             cols = c("grey90", bms_color$doublet[["doublet"]]), -->
<!--             min.cutoff = .5, -->
<!--             combine = FALSE) -->
<!-- ``` -->

<!-- ```{r Doublet automatic call on UMAP split by sample, fig.width=12, fig.height=6} -->
<!-- DimPlot(gex, group.by = "scDblFinder_class_0.002", -->
<!--         cols = bms_color$doublet, -->
<!--             split.by = "orig.ident", ncol = 6, -->
<!--             combine = FALSE)  -->
<!-- ``` -->
<!-- ## Histogram of doublet scores -->

<!-- ```{r hist of doublet scores} -->
<!-- thresholds <- c(0.001, 0.002, 0.005, 0.01) -->

<!-- hist(gex$scDblFinder_score_0.001) -->
<!-- hist(gex$scDblFinder_score_0.002) -->
<!-- hist(gex$scDblFinder_score_0.005) -->
<!-- hist(gex$scDblFinder_score_0.01) -->

<!-- plots <- RidgePlot(gex, -->
<!--           features = sapply(thresholds, function (x) {paste0("scDblFinder_score_", as.character(x))}), -->
<!--           group.by = "orig.ident", -->
<!--           cols = bms_color$sample, -->
<!--           combine = F)  -->

<!-- # Add scale_x_continuous to each plot -->
<!-- plots <- lapply(plots, function(p) { -->
<!--   p + scale_x_continuous(breaks = seq(0, 1, 0.1)) + guides(fill="none") -->
<!-- }) -->

<!-- plots -->
<!-- ``` -->

<!-- ```{r Violin plot of automatic doublet calls, fig.width=8} -->
<!-- p <- VlnPlot(gex, "scDblFinder_score_0.002",  -->
<!--         group.by = "orig.ident", split.by = "scDblFinder_class_0.002", split.plot = T, -->
<!--         pt.size = 0, add.noise = T) -->
<!-- p -->
<!-- p + guides(fill = "none") +  -->
<!--     scale_y_continuous(breaks = seq(0.8, 1, 0.02), limits = c(.8,1)) + -->
<!--     geom_hline(yintercept = seq(.94, 1, .01), linetype = "dotted") -->
<!-- ``` -->

<!-- ## Set manual doublet thresholds -->

<!-- ```{r Sample-specific thresholds} -->
<!-- # Define thresholds for each sample -->
<!-- # using the plot above -->
<!-- sample_thresholds <- list( -->
<!--     "YK01_En7Apos" = .98, -->
<!--     "YK02_En7Bneg" = .99, -->
<!--     "YK03_En7Bpos" = .98, -->
<!--     "YK04_En7Cneg" = .97, -->
<!--     "YK05_En7Cpos" = .96, -->
<!--     "YK06_En7ReAneg" = .97, -->
<!--     "YK07_En7ReApos" = .97, -->
<!--     "YK08_En7ReBpos" = .99, -->
<!--     "YK09_En28Apos" = .98, -->
<!--     "YK10_En28Cneg" = .99, -->
<!--     "YK11_En28Cpos" = .99, -->
<!--     "YK12_En28Dneg" = .98, -->
<!--     "YK13_En28Dpos" = .98, -->
<!--     "YK14_En28ReBneg" = .97, -->
<!--     "YK15_En28ReBpos" = .98, -->
<!--     "YK16_En28ReCneg" = .97, -->
<!--     "YK17_En28ReCpos" = .98 -->
<!-- ) -->

<!-- # Initialize a vector to store doublet classifications -->
<!-- gex$scDblFinder_class <- NA  # Placeholder for classification ("doublet" or "singlet") -->

<!-- # Iterate through samples and apply thresholds -->
<!-- for (sample in names(sample_thresholds)) { -->
<!--   # Get the threshold for the current sample -->
<!--   threshold <- sample_thresholds[[sample]] -->

<!--   # Identify cells belonging to the current sample -->
<!--   sample_cells <- which(gex$orig.ident == sample) -->

<!--   # Classify cells as "doublet" or "singlet" based on the threshold -->
<!--   gex$scDblFinder_class[sample_cells] <- ifelse( -->
<!--     gex$scDblFinder_score_0.002[sample_cells] > threshold,  -->
<!--     "doublet",  -->
<!--     "singlet" -->
<!--   ) -->
<!-- } -->

<!-- # Check the classification results -->
<!-- doublet_clasification <- table(gex$scDblFinder_class) -->
<!-- print(doublet_clasification) -->
<!-- sprintf("%.1f%% of all cells called.",  -->
<!--         100 * doublet_clasification[["doublet"]] / length(Cells(gex))) %>% cat() -->
<!-- ``` -->
<!-- Doublet threshold was set to match the estimated ratio from 10x manual and cell numbers. -->

<!-- ## Visualize manual doublet call results -->

<!-- ```{r Violin plot of manual doublet calls, fig.width=8} -->
<!-- p <- VlnPlot(gex, "scDblFinder_score_0.002",  -->
<!--         group.by = "orig.ident", split.by = "scDblFinder_class", split.plot = T, -->
<!--         pt.size = 0, add.noise = T) -->
<!-- p -->
<!-- p + guides(fill = "none") + -->
<!--     scale_y_continuous(breaks = seq(0.7, 1, 0.02), limits = c(.7,1)) + -->
<!--     geom_hline(yintercept = seq(.94, 1, .01), linetype = "dotted") -->
<!-- ``` -->

<!-- ```{r UMAP with manual doublet calls, fig.width=3, fig.height=2.5} -->
<!-- DimPlot(gex, group.by = "scDblFinder_class", -->
<!--         cols = bms_color$doublet) -->
<!-- ``` -->
<!-- ## Mixing with Seurat clusters -->

<!-- ```{r Seurat clusters sorted by doublet proportion} -->
<!-- prop_table <- proportions(table(gex$scDblFinder_class, gex$seurat_clusters), 2) -->
<!-- doublet_ratio <- sort(prop_table["doublet",]) -->
<!-- print(doublet_ratio) -->
<!-- barplot(doublet_ratio, #legend.text = c("doublet", "singlet"), -->
<!--         xlab = "Doublet proportion", ylab = "Seurat clusters", horiz = T, las = 2) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- table(gex$scDblFinder_class, gex$seurat_clusters) -->
<!-- ``` -->

<!-- ## Properties of the called doublets -->

<!-- ```{r} -->
<!-- VlnPlot(gex, "nCount_RNA",  -->
<!--         split.by = "scDblFinder_class", group.by = "seurat_clusters", -->
<!--         pt.size = 0, add.noise = T, split.plot = T)  -->

<!-- ``` -->


<!-- ```{r Aberrant coexpression} -->
<!-- # Define the genes of interest -->
<!-- excitatory_genes <- c("Slc17a7", "Slc17a6")  # Markers for excitatory neurons -->
<!-- inhibitory_genes <- c("Gad1", "Gad2")        # Markers for inhibitory neurons -->

<!-- # Identify cells expressing each gene set -->
<!-- excitatory_cells <- colSums(GetAssayData(gex, slot = "data")[excitatory_genes, , drop = FALSE] > 0) > 0 -->
<!-- inhibitory_cells <- colSums(GetAssayData(gex, slot = "data")[inhibitory_genes, , drop = FALSE] > 0) > 0 -->

<!-- # Find cells expressing BOTH excitatory and inhibitory markers -->
<!-- double_positive_cells <- which(excitatory_cells & inhibitory_cells) -->

<!-- # Check the classification of these cells (doublet or singlet) -->
<!-- double_positive_classifications <- table(gex$scDblFinder_class[double_positive_cells]) -->

<!-- # Print results -->
<!-- cat("Number of cells expressing both excitatory and inhibitory markers:\n") -->
<!-- cat(sprintf("%d / %d (%.1f%%)",  -->
<!--             length(double_positive_cells), length(Cells(gex)), -->
<!--             100 * length(double_positive_cells) / length(Cells(gex)))) -->

<!-- cat("\n\n\nClassification of these cells:\n") -->
<!-- print(double_positive_classifications) -->

<!-- cat("\nTotal classification counts:\n") -->
<!-- total_classifications <- table(gex$scDblFinder_class) -->
<!-- print(total_classifications) -->

<!-- cat("\nProportions of doublets and singlets:\n") -->
<!-- cat(sprintf("%.1f%% of doublets, %.1f%% of singlets", -->
<!--     double_positive_classifications[["doublet"]] / total_classifications[["doublet"]] * 100, -->
<!--     double_positive_classifications[["singlet"]] / total_classifications[["singlet"]] * 100)) -->

<!-- # Optionally, extract the barcodes of these cells for further analysis -->
<!-- double_positive_barcodes <- colnames(gex)[double_positive_cells] -->

<!-- ``` -->


<!-- # Save to output -->

<!-- Temporarily only using the singlets called from the GEX modality: -->

<!-- ```{r} -->
<!-- Sys.setenv(HDF5_USE_FILE_LOCKING=FALSE, RHDF5_USE_FILE_LOCKING=FALSE) -->
<!-- ``` -->


<!-- ```{r Save Seurat object after downsampling and doublet detection} -->
<!-- saveRDS(gex, gex.obj.path) -->
<!-- ``` -->



# Session info

```{r sinfo, cache = FALSE}
sessionInfo()
```
