---
title: "BMS"
author: "Yuxi Ke"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    theme: "flatly"
    number_sections: yes
    toc: yes
    toc_float: yes
    code_folding: hide
---

```{r config, warning = FALSE}

# **MODIFY THIS CHUNK**
project_id <- "bms" # determines the name of the cache folder
doc_id     <- file.path("03-3a_dataset_annotate") # determines name of the subfolder of `figures` where pngs/pdfs are saved
dataout        <- file.path(here::here("Output/"), doc_id); dir.create(dataout, recursive = TRUE)
figout     <- file.path(here::here("Fig/"), doc_id); dir.create(figout, recursive = TRUE)
```

```{r setup, include = FALSE}

# NO NEED TO MODIFY THIS CHUNK
here::i_am(".gitignore")
knitr::opts_chunk$set(message = TRUE,
                      warning = FALSE,
                      error = FALSE,
                      cache = FALSE,
                      fig.path = figout,
                      fig.keep = "all",
                      dev = c("png", "pdf"),
                      cache.lazy = FALSE)

grDevices::pdf.options(useDingbats = FALSE)
options(knitr.table.format = "html") 

reload <- function() {
    devtools::load_all("~/workspace/bms")
}

redoc <- function() {
    devtools::document("~/workspace/bms")
}

library(here)
library(Seurat)
library(ArchR)
library(dbscan)
library(tidyr)
library(ggplot2)
library(dplyr)

reload()
bms::setup_archr(); bms::set_theme()
set.seed(100)
```

***

# Overview

Here, I annotated the clusters, filtered out two dublet clusters found during 
annotation, and added sample metadata.

Input:

    - An ArchR project
    - A Seurat object
    - Seurat cluster annotation csv files
    - A sample metadata csv file
    
Output:

    - Clean data with metadata, including small clusters
        - An ArchR project (original dir)
        - A Seurat object (original dir)
    - Clean data with metadata, leaving only big clusters for differential analysis
        - An ArchR projct (new dir)
        - A Seurat object (new dir)
    
Yuxi Ke, 01/22/2025


# Load data

## Define file names

```{r, include=TRUE}
location <- "local" # {"local", "sherlock"}

if (location == "local") {
    archr.proj.path <- here::here("Data/ArchRProjects/BMS_clean")
    gex.obj.path <- here::here("Data/SeuratObjects/gex_clean.rds")
    annotation.path <- here::here("Data/Metadata/seurat_clusters_annotation.tsv")
    annotation2.path <- here::here("Data/Metadata/seurat_clusters_annotation_hierarchy.tsv")
    
    new.archr.proj.path <- here::here("Data/ArchRProjects/BMS_diff")
    new.gex.obj.path <- here::here("Data/SeuratObjects/BMS_diff.rds")
} else {
    oak.path <- "/oak/stanford/groups/wjg/kyx/data/bms"
    archr.proj.path <- file.path(oak.path, "archr/BMS_mm10")
    gex.obj.path <- file.path(oak.path, "seurat/BMS_clean.rds")
}
```

## Read data

```{r}
proj <- loadArchRProject(archr.proj.path, showLogo = FALSE)
gex <- readRDS(gex.obj.path)
sample.metadata <- bms::load_sample_metadata(here::here("Data/Metadata/YK_sample_sheet.csv"))

# Set index
rownames(sample.metadata) <- sample.metadata$CombinedID
# Remove repeat column
sample.metadata <- sample.metadata %>% dplyr::select(-sample_id)
cls_anno <- read.delim(annotation.path)
cls_hier_anno <- read.delim(annotation2.path)
```

# Make sense of the clusters

```{r BMS_clean UMAP}
Idents(gex) <- "RNA_snn_res.1"
p <- list()
label2plot <- c("subclass", "neighborhood", "class", "seurat_clusters", "archr_clusters")
for (label.col in label2plot) {
    n <- nrow(unique(gex[[label.col]]))
    p[[label.col]] <- DimPlot(gex, group.by = label.col, label = T,
                              cols = bms::get_color_map(bms::cmaps$circus, n + 1))
}
p
```

```{r Cluster sizes}
table(gex$seurat_clusters)
```
```{r}
DotPlot(
    gex,
    c("Fezf2", "Pou3f1", "Bcl6", "Bcl11b", "Etv1",
      "Tle4", "Foxp2", "Rasgrf2"),
    idents = c("6", "14", "15","27")
)
```


```{r Check stress markers}
oxidative_stress_markers <- c(
  "Hspa1a", "Hspa1b", "Gclc", "Gclm", "Nfe2l2", "Hmox1", 
  "Prdx1", "Prdx2", "Sod1", "Sod2", "Gpx1", "Gpx4"
)

er_stress_markers <- c(
  "Atf4", "Ddit3", "Xbp1", "Hspa5", "Eif2ak3", 
  "Hyou1", "Derl1", "Derl3"
)

heat_shock_markers <- c(
  "Hsp90aa1", "Hsp90ab1", "Dnajb1", "Dnajc3", 
  "Bag3", "Cryab"
)

dna_damage_markers <- c(
  "Trp53", "Gadd45a", "Rpa1", "H2ax"
)

inflammatory_stress_markers <- c(
  "Il6", "Il1b", "Tnf", "Cxcl10", "Ccl2", 
  "Nfkb1", "Rela", "Nos2", "Ptgs2"
)

apoptosis_markers <- c(
  "Bax", "Bak1", "Bcl2", "Bcl2l1", 
  "Caspase3", "Caspase7", "Cycs"
)


DotPlot(gex,
        assay = "RNA",
        features = apoptosis_markers,
        idents = c("1", "23", "2", "5", "0", "14")) + NoLegend() + RotatedAxis()
```


```{r Highlight a subset of seurat clusters}
# # CA2
# DimPlot(gex, #cols = bms::pal$seurat_clusters,
#         label = T, shuffle = T,
#         cells.highlight = WhichCells(gex, idents = c("0", "3", "17", "25"))) + ggtitle("CA2")
```


# Filter extra identified doublet clusters

```{r}
dbl_clusters <- c("26", "28")
gex <- subset(gex, subset = seurat_clusters %ni% dbl_clusters)
proj <- proj[Cells(gex),]
```




# Add cluster annotation

```{r Format and print annotation}
cls_anno$seurat_clusters <- as.character(cls_anno$seurat_clusters)
cls_anno
```

```{r}
# Add `Ident` as a new metadata column in `gex` under `cell_cluster`
gex@meta.data$cell_cluster <- cls_anno$Ident[match(gex@meta.data$seurat_clusters, cls_anno$seurat_clusters)]
# Check there's no NA
stopifnot(! any(is.na(gex@meta.data$cell_cluster)))

# Set the new cell identities for the Seurat object
Idents(gex) <- gex@meta.data$cell_cluster

stopifnot(Cells(gex) == proj$cellNames)

proj <- ArchR::addCellColData(proj, 
                              data = gex@meta.data$cell_cluster,
                              cells = Cells(gex),
                              name = "cell_cluster",
                              force = T)
# Check no NA
stopifnot(!any(is.na(proj$cell_cluster)))
```

```{r}
DimPlot(gex, group.by = "cell_cluster", label = T, repel = T,
        cols = bms::pal$cell_type)
plotEmbedding(proj, name = "cell_cluster", labelSize = 0)
```
```{r}
ggplot(gex[[]], aes(fill=subclass, x=cell_cluster)) + 
    geom_bar(position = "fill") + coord_flip() + scale_fill_discrete(type=bms::get_color_map(bms::cmaps$kelly, 26))
```

# Attach sample metadata

## to ATAC

```{r Attach sample metadata to ArchR}
# Extract cellColData from the ArchR project
cell_col_data <- getCellColData(proj)

# Create a dataframe with cell IDs and Sample
cell_sample_data <- data.frame(
  cellID = rownames(cell_col_data),
  Sample = cell_col_data$Sample,
  stringsAsFactors = FALSE
)

# Choose metadata columns to add
metadata.cols <- c("CombinedID", "SampleID", "Group", "Gating", "Condition", "Time", "Recall", "AgeWeeks", "Gender")
# Perform the left join with selected columns from sample.metadata
joined_metadata <- cell_sample_data %>%
  left_join(sample.metadata[, metadata.cols], 
            by = c("Sample" = "CombinedID"))

# Add selected metadata columns back to the ArchR project
for (col in metadata.cols[metadata.cols != "CombinedID"]) {
  metadata_vector <- joined_metadata[[col]]
  names(metadata_vector) <- joined_metadata$cellID
  
  proj <- addCellColData(
    ArchRProj = proj,
    data = as.character(metadata_vector),
    name = col,
    cells = joined_metadata$cellID,
    force = TRUE
  )
}
```

## To GEX

```{r Add metadata to gex}
stopifnot(joined_metadata$cellID == Cells(gex))

for (col in metadata.cols[metadata.cols != "CombinedID"]) {
  metadata_vector <- joined_metadata[[col]]
  # names(metadata_vector) <- joined_metadata$cellID
  
  gex <- AddMetaData(
      gex,
      metadata = as.character(metadata_vector),
      col.name = col
  )
}
```


# Save diff only to a new path

## Subset GEX

```{r Identify clusters to remove from diff}
n_cell <- table(gex$cell_cluster)
print(n_cell)
small_cell_clusters <- n_cell[n_cell < 200] %>% names()
print(small_cell_clusters)
not_diff_cell_clusters <- cls_hier_anno[!cls_hier_anno$in_diff,]$cell_cluster
print(not_diff_cell_clusters)
```

```{r Subset to diff}
gex_diff <- subset(gex, subset = cell_cluster %ni% union(small_cell_clusters, not_diff_cell_clusters))
```

## Attach hierarchical cell type annotation


```{r Add hierarchical cell type annotation}
print(cls_hier_anno)

gex_diff@meta.data$cell_type <- cls_hier_anno$cell_type[match(gex_diff@meta.data$cell_cluster, 
                                                cls_hier_anno$cell_cluster)]
stopifnot(!any(is.na(gex_diff$cell_type)))

gex_diff@meta.data$cell_neighborhood <- cls_hier_anno$cell_neighborhood[match(gex_diff@meta.data$cell_cluster, 
                                                cls_hier_anno$cell_cluster)]
gex_diff@meta.data$cell_class <- cls_hier_anno$cell_class[match(gex_diff@meta.data$cell_cluster, 
                                                cls_hier_anno$cell_cluster)]
```

## Reset levels

```{r Reset gex metadata levels}
cols <- colnames(cls_hier_anno)[1:4]
for (col in cols){
    metadata <- factor(x = gex_diff@meta.data[,col], levels = unique(cls_hier_anno[[col]]))
    gex_diff <- AddMetaData(gex_diff, metadata, col.name = col)
}

Idents(gex_diff) <- gex_diff$cell_type
```


```{r Re-UMAP after subsetting to diff only}
gex_diff <- RunUMAP(gex_diff, dims = 1:40)
```

```{r Final GEX UMAPs}
DimPlot(gex_diff, group.by = "cell_cluster",
        label = T, cols = bms::pal$cell_type)
DimPlot(gex_diff, group.by = "cell_type",
        label = T, cols = bms::pal$cell_type)
DimPlot(gex_diff, group.by = "cell_neighborhood",
        label = T, cols = bms::pal$cell_type)
DimPlot(gex_diff, group.by = "cell_class",
        label = T, cols = bms::pal$cell_type)
DimPlot(gex_diff, group.by = "seurat_clusters",
        label = T, cols = bms::pal$seurat_clusters)
```

## Subset ATAC

Subset and attach cell annotation to ArchR project.

```{r Subset BMS_diff ArchR project}
proj_diff <- proj[Cells(gex_diff),]

stopifnot(Cells(gex_diff) == proj_diff$cellNames)

for (lbl in c("cell_type", "cell_neighborhood", "cell_class")){
proj_diff <- ArchR::addCellColData(proj_diff, 
                              data = as.character(gex_diff@meta.data[[lbl]]),
                              cells = Cells(gex_diff),
                              name = lbl,
                              force = T)
}

proj_diff <- addUMAP(proj_diff, force = T)
plotEmbedding(proj_diff, name = "cell_cluster")
plotEmbedding(proj_diff, name = "cell_type")
plotEmbedding(proj_diff, name = "cell_neighborhood")
plotEmbedding(proj_diff, name = "cell_class")

proj_diff <- addImputeWeights(proj_diff)
```

## Remove extra columns

```{r Remove repeated or unused Seurat clusters}
cols2rm <- c(
    "RNA_snn_res.2", "RNA_snn_res.0.3", "RNA_snn_res.0.6", "RNA_snn_res.5", "RNA_snn_res.3",
    "class.1", "subclass.1", "neighborhood.1",
    "class.prediction.score.max.2", "subclass.prediction.score.max.2", "neighborhood.prediction.score.max.2",
    "neighborhood.prediction.score.max", "class.prediction.score.max", "subclass.prediction.score.max", "cluster.prediction.score.max",
    "is_neuron", "scDblFinder_class",
    "class", "subclass", "neighborhood", "cluster"
)
for (col in cols2rm) {
    gex_diff@meta.data[[col]] <- NULL
    
}
```


## Remove YK11

```{r Show cell types in YK11}
tmp <- subset(gex_diff, subset = orig.ident == "YK11_En28Cpos")
table(Idents(tmp))
```


```{r Remove YK11_En28Cpos}
gex_diff <- subset(gex_diff, subset = orig.ident != "YK11_En28Cpos")
idxSample <- BiocGenerics::which(proj_diff$Sample %ni% "YK11_En28Cpos")
proj_diff <- subsetCells(proj_diff, proj_diff$cellNames[idxSample])
```

## Write to disk

```{r Save to disk}
saveRDS(gex_diff, new.gex.obj.path)
ArchR::saveArchRProject(proj_diff, new.archr.proj.path)
```

# Gene ranges
```{r Gene ranges}
library(rtracklayer)
library(GenomeInfoDb)

############
use_saved_genes_matched = TRUE
############

if (use_saved_genes_matched) {
    genes_matched <- readRDS(here::here("Data/Metadata/gene_ranges.rds"))
} else {
    gtf_url <- "ftp://ftp.ensembl.org/pub/release-102/gtf/mus_musculus/Mus_musculus.GRCm38.102.gtf.gz"
    gtf_file <- here::here("/Data/Downloads/Mus_musculus.GRCm38.102.gtf.gz")
    download.file(gtf_url, destfile = gtf_file)
    
    gtf <- import(gtf_file)
    genes_gtf <- gtf[gtf$type == "gene"]
    
    rm(gtf)
    gene_symbols <- mcols(genes_gtf)$gene_name
    names(genes_gtf) <- gene_symbols
    
    symbols <- rownames(gex_diff)
    common <- intersect(symbols, gene_symbols)
    
    gex_diff <- gex_diff[common,]
    genes_matched <- genes_gtf[common]
    genes_matched <- genes_matched[match(rownames(gex_diff), names(genes_matched))]
    
    seqlevelsStyle(genes_matched) <- "UCSC"
    standard_chr <- paste0("chr", c(1:19))
    genes_matched <- keepSeqlevels(genes_matched, standard_chr, pruning.mode = "coarse")
    
    saveRDS(genes_matched, here::here("Data/Metadata/gene_ranges.rds"))
}

gex_diff <- gex_diff[names(genes_matched),]

saveRDS(gex_diff, here::here("Data/SeuratObjects/gex_gr.rds"))
```

# Session info

```{r sinfo, cache = FALSE}
sessionInfo()
```
